### ISSUE TITLE
forbidden identity

### ISSUE THREAD

**Issue by donileo** [aws/aws-sdk-ios#165]
I've been investigating a super weird error that occurs when my app that was previously log'd in with Facebook then is deleted and reinstalled. When the app asks to register for push notifications Amazons SDK will give a User identity is forbidden. Here is the error:

```
AWSiOSSDKv2 [Verbose] AWSURLRequestSerialization.m line:110 | -[AWSJSONRequestSerializer serializeRequest:headers:parameters:] | Request body: [{"IdentityId":"us-east-1:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxx"}]
2015-03-18 23:00:56.554 Rendxxxxxx[2135:1168595] AWSiOSSDKv2 [Debug] AWSURLResponseSerialization.m line:85 | -[AWSJSONResponseSerializer responseObjectForResponse:originalRequest:currentRequest:data:error:] | Response header: [{
    "Content-Length" = 129;
    "Content-Type" = "application/x-amz-json-1.1";
    Date = "Thu, 19 Mar 2015 03:00:55 GMT";
    nnCoection = close;
    "x-amzn-RequestId" = "29c5be36-cde4-11e4-b63f-217feac9a8d2";
}]
2015-03-18 23:00:56.554 Rendxxxxxx[2135:1168595] AWSiOSSDKv2 [Verbose] AWSURLResponseSerialization.m line:90 | -[AWSJSONResponseSerializer responseObjectForResponse:originalRequest:currentRequest:data:error:] | Response body: [{"__type":"NotAuthorizedException","message":"Access to Identity 'us-east-1:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxx' is forbidden."}]
2015-03-18 23:00:56.557 Rendxxxxxx[2135:1168595] AWSiOSSDKv2 [Error] AWSIdentityProvider.m line:214 | __45-[AWSAbstractCognitoIdentityProvider refresh]_block_invoke_2 | GetOpenIdToken failed. Error is [Error Domain=com.amazonaws.AWSCognitoIdentityErrorDomain Code=9 "The operation couldn’t be completed. (com.amazonaws.AWSCognitoIdentityErrorDomain error 9.)" UserInfo=0x17086ca00 {__type=NotAuthorizedException, message=Access to Identity 'us-east-1:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxx' is forbidden.}]
2015-03-18 23:00:56.558 Rendxxxxxx[2135:1168595] AWSiOSSDKv2 [Error] AWSCredentialsProvider.m line:581 | __40-[AWSCognitoCredentialsProvider refresh]_block_invoke354 | Unable to refresh. Error is [Error Domain=com.amazonaws.AWSCognitoIdentityErrorDomain Code=9 "The operation couldn’t be completed. (com.amazonaws.AWSCognitoIdentityErrorDomain error 9.)" UserInfo=0x17086ca00 {__type=NotAuthorizedException, message=Access to Identity 'us-east-1:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxx' is forbidden.}]
```

Now deep diving into amazons sdks i find a peculiar area of code that look a bit fishy to me:

lines 405-412 of AWSCredentialsProvider.m

```
// If the identity provider has an identity id, use it
        if (identityProvider.identityId) {
            _keychain[AWSCredentialsProviderKeychainIdentityId] = identityProvider.identityId;
        }
        // Otherwise push whatever is in the keychain down to the identity provider
        else {
            identityProvider.identityId = _keychain[AWSCredentialsProviderKeychainIdentityId];
        }
```

which states that if no identity_id is found.. use whatever was already in the keychain, but remember the keychain persists even if the app is deleted. So to better explain, if we logged in previously with Facebook the previous identity_id gets stored  in the keychain. Then when the app gets removed from the device and reinstalled the above code picks up the stored auth'ed identity_id from the keychain from before and uses it in our call to GetOpenTokenID. This seems bad because now any calls to AmazonAWS will try to access this authenticated_id which we can only access after the device gets a new token from Facebook authenticating with that previous login. If there is no Facebook token (if user doesn't do a Facebook login) this will result in an invalid request to GetOpenTokenID with the identity_id set to the auth_ed id from before but no logins dictionary.. which I believe is the error above. Hopefully I did proper analysis of this, if this is indeed a bug please fix asap!

after looking around a bit i also found this https://github.com/aws/aws-sdk-ios/issues/93.. in which you guys state its not reproducible. in my case it is very reproducible. Simply add aws-sdk/cognito. perform a login with Facebook, delete the app, reinstall the app then try to make a call to amazon sns for example to add push notifications for guests.

As a temporary solution executing the clearKeychain method on the AWSCredentials will fix the problem asap. but that would mean somehow being able to catch your app was reinstalled to run this method.

**Comment by yosuke-matsuda** [comment]
This seems like a bug in the SDK. We will investigate and try to address it. I will update the thread once the fix becomes available. Thanks.

**Comment by ramanakanth** [comment]
i am experiencing the same issue . is there any update on the bug

**Comment by turacma** [comment]
There's been some confusion about this issue. It's currently operating as designed. If you are receiving the 'Access to Identity 'us-east-1:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxx' is forbidden.' it because the value being cached in the keychain is associated with a authenticated identity and you are not providing valid credentials.

As this is how the class is designed, I'd be interested to here how the OP @donileo (and @ramanakanth as well), would expect this to behave.

**Comment by ramanakanth** [comment]
yeah i see that issue with keychain. for timebeing the workaround fix i made was calling [_credentialsprovider clearkeychain] if getting this error. This way atleast i can avoid error but user has to login again.
Thanks

**Comment by donileo** [comment]
the problem lies in that the keychain the AWSSDK is persisting the data to persists even if the app is deleted. So even i did have valid credentials right now and all is working when the app is deleted/reinstalled the AWSSDK will reused the cached credentials from the app before it was deleted. Hence resulting in the forbidden identity. I don't think this should work as designed this way. Just because one user had valid credentials doesn't mean the AWSSDK should automatically use the keychain data all the time. Persisting this information as a cache to the iOS keychain seems wrong imo.

**Comment by turacma** [comment]
The app's document directory is not secure as the keychain, hence our choice in using that method of storage. I appreciate that this is a pain point, but obviously we don't want to reduce the security profile of the data we're storing there.

Perhaps a compromise is to use the app's document directory (or NSUserDefaults) to detect the presence of cached data, but to continue to actually store that data in the keychain.

**Comment by donileo** [comment]
That sounds like a good compromise. My initial suggestion to store the data on the document directory was based on the assumption that the data didn't need to be secured which is a bad assumption :), I mean why would you be using the keychain anyways.... But yes checking for a token on the app directory which then dictates whether to reuse the existing keychain identity or clear it and start a new one sounds like a possible solution. I think the initial issue is that somehow you guys believed the Keychain would be deleted upon an app deletion, I thought the same initially but nope unfortunately its not :). After a bit more thinking though, why not just store this cached token (not the keychain data) on AWS itself.

**Comment by turacma** [comment]
One of the core features of Cognito is for bootstrapping credentials to access AWS. Storing the cached values on AWS would create a catch-22.

**Comment by donileo** [comment]
Store it in the app directory it is then :). However it definitely should be done in a way that even if the app directory is cleared the library will function. So if app directory is cleared (token is missing) then simply clear the identities and let them be rebuilt. Right now I can run clearKeychain 10x on the AWSCredentials and I will not get the forbidden identity access error, it'll just work because I am assuming the library just populates itself correctly after that.

**Comment by vsochinsky** [comment]
Also experiencing this exact problem as OP, is this thing being fixed or fixed already?

**Comment by GreyEcologist** [comment]
running into the same issue, would love a best practice on the topic, @turacma

**Comment by ajsharp** [comment]
@GreyEcologist Agreed, a recommendation on how to deal with this would be very helpful.

**Comment by ajsharp** [comment]
The issue I'm having is that there appears to be cached credentials somewhere, and unless is I call `[credentialsProvider refresh]` manually I get an "invalid login token" error. Ideally, it would be very helpful if we could get some instruction / documentation on how to permanently overwrite old login credentials, or at least refresh what's there. The existing SDK documentation says that you should never have to call `refresh`, so either there's a bug in the SDK, or the docs are misleading.

**Comment by SunnyGurnani** [comment]
Any update on this more then a year now. clearKeyChain is a very BAD workaround because it clears all Cognito Ids and It ends up creating New Cognito For Unauthenticated Users.

So Lets say a User Logins and Logouts 5 times on a Mobile Phone there will be total 6 Cognito Id's created.  Every Logout creates new CognitoId because of clearKeyChain

**Comment by yosuke-matsuda** [comment]
We've released the AWS Mobile SDK for iOS 2.4.0, which addresses the issue. Thanks.

**Comment by donileo** [comment]
That definitely took some time.. but great to hear :)
